src/Core/src/Hosting/MauiServiceCollection.cs

```diff
@@ -1,110 +1,76 @@
 using System;
 using System.Collections;
 using System.Collections.Generic;
-using System.Linq;
 using Microsoft.Extensions.DependencyInjection;
 
 namespace Microsoft.Maui.Hosting
 {
 	class MauiServiceCollection : IMauiServiceCollection
 	{
-		static string s_error => "This Collection is based on a non ordered Dictionary";
-
-		readonly Dictionary<Type, Func<IServiceProvider, object?>?> _factories = new Dictionary<Type, Func<IServiceProvider, object?>?>();
 		readonly List<ServiceDescriptor> _descriptors = new List<ServiceDescriptor>();
+		readonly Dictionary<Type, ServiceDescriptor> _descriptorDictionary = new Dictionary<Type, ServiceDescriptor>();
 
 		public int Count => _descriptors.Count;
 
 		public bool IsReadOnly => false;
 
-		public ICollection<Type> Keys => _factories.Keys;
-
-		public ICollection<Func<IServiceProvider, object?>?> Values => _factories.Values;
-
-		public Func<IServiceProvider, object?>? this[Type key] { get => _factories[key]; set => _factories[key] = value; }
+		public ServiceDescriptor this[int index]
+		{
+			get => _descriptors[index];
+			set => _descriptors[index] = value;
+		}
 
 		public void Add(ServiceDescriptor item)
 		{
-			if (!_descriptors.Contains(item))
-				_descriptors.Add(item);
-			if (item.ImplementationType != null)
-				_factories[item.ServiceType] = new Func<IServiceProvider, object?>(provider => Activator.CreateInstance(item.ImplementationType));
-			else if (item.ImplementationInstance != null)
-				_factories[item.ServiceType] = new Func<IServiceProvider, object?>(provider => item.ImplementationInstance);
-			else if (item.ImplementationFactory != null)
-				_factories[item.ServiceType] = item.ImplementationFactory;
-			else
-				throw new InvalidOperationException($"You need to provide an {nameof(item.ImplementationType)} or a {nameof(item.ImplementationInstance)} ");
+			if (_descriptors.Contains(item))
+				return;
```

Comment:
I m not sure if we should throw or log something here to say to the user we don't support adding several implementations of the same type.

Reply:
Should we throw or maybe override existing? So "last one wins"

Reply:
we need to override, lets say Maui register ButtonHandler for IButton, and the user wants to register a CustomButtonHandler for IButton. the last one should win.

Reply:
It actually is already the last one wins. if that exact instance of the descriptor is in already, it skips. If not, it overrides the service in the dictionary.

In actual fact, the descriptors list could have double registrations, but the dictionary always is the last one.